File: app/src/main/java/com/yfbx/widgets/util/BitmapUtil.java
package com.yfbx.widgets.util;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Environment;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

/**
 * Author: Edward
 * Date: 2019/1/7
 * Description:
 */


public class BitmapUtil {

    /**
     * 将 Bitmap 保存到SD卡
     */
    public static void saveBitmap(Bitmap bitmap, String fileName) {
        String sd = Environment.getExternalStorageDirectory().getPath();
        File file = new File(sd, fileName + ".png");
        try {
            FileOutputStream outStream = new FileOutputStream(file);
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, outStream);
            outStream.flush();
            outStream.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * 通知媒体扫描器,扫描媒体文件
     */
    public static void notifyMediaScanner(Context context, File file) {
        Intent intent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
        Uri uri = Uri.fromFile(file);
        intent.setData(uri);
        context.sendBroadcast(intent);
    }
}

File: app/src/main/java/com/yfbx/widgets/util/BitmapUtil.kt
package com.yfbx.widgets.util

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.media.MediaScannerConnection
import android.os.Environment
import android.view.View
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

/**
 * Author: Edward
 * Date: 2019/1/7
 * Description:
 */

/**
 * 保存 Bitmap
 */
fun Bitmap.save(context: Context): Boolean {
    var isSuccess = false
    val path = Environment.DIRECTORY_PICTURES + "/" + System.currentTimeMillis() + ".png"
    val file = File(Environment.getExternalStorageDirectory(), path)
    try {
        val fos = FileOutputStream(file)
        compress(Bitmap.CompressFormat.PNG, 100, fos)
        fos.flush()
        fos.close()
        isSuccess = true
        recycle()
    } catch (e: IOException) {
        e.printStackTrace()
    }
    MediaScannerConnection.scanFile(context, arrayOf(file.path), null, null)
    return isSuccess
}

/**
 * 保存Drawable
 */
fun Drawable.save(context: Context): Boolean {
    val bitDrawable = this as BitmapDrawable
    val bitmap = bitDrawable.bitmap
    return bitmap.save(context)
}


/**
 * View 截图
 */
fun View.capture(): Bitmap {
    val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
    bitmap.eraseColor(Color.TRANSPARENT)
    draw(Canvas(bitmap))
    return bitmap
}


/**
 * 保存View 截图
 */
fun View.saveCapture(): Boolean {
    return capture().save(context)
}
